lib/WP_Auth0_LoginManager.php

```diff
@@ -297,47 +294,60 @@ public function redirect_login() {
    * @throws WP_Auth0_LoginFlowValidationException
    *
    * @see https://auth0.com/docs/api-auth/tutorials/implicit-grant
+   * @see /wp-content/plugins/auth0/assets/js/implicit-login.js
    */
   public function implicit_login() {
 
+    // Posted from the login page here
     $token = $_POST['token'];
-    $state_decoded = $this->get_state( TRUE );
-
-    $secret = $this->a0_options->get_client_secret_as_key();
 
     try {
-      // Decode the user
-      $decodedToken = JWT::decode( $token, $secret, array(  $this->a0_options->get_client_signing_algorithm() ) );
+      $decodedToken = JWT::decode(
+        $token,
+        $this->a0_options->get_client_secret_as_key(),
+        array(  $this->a0_options->get_client_signing_algorithm() )
+      );
+    } catch( UnexpectedValueException $e ) {
+      WP_Auth0_ErrorManager::insert_auth0_error( __METHOD__, $e );
+      error_log( $e->getMessage() );
+      throw new WP_Auth0_LoginFlowValidationException();
+    }
 
-      // validate that this JWT was made for us
-      if ( $this->a0_options->get( 'client_id' ) !== $decodedToken->aud ) {
-        throw new WP_Auth0_LoginFlowValidationException( 'This token is not intended for us.' );
+    // Validate that this JWT was made for us
+    if ( $this->a0_options->get( 'client_id' ) !== $decodedToken->aud ) {
+      throw new WP_Auth0_LoginFlowValidationException(
+        __( 'This token is not intended for us', 'wp-auth0' )
+      );
+    }
+
+    // Validate the nonce if one was included in the request (auto-login)
+    if ( $this->a0_options->get( 'auto_login' ) ) {
+      $nonce = isset( $decodedToken->nonce ) ? $decodedToken->nonce : null;
+      if ( ! WP_Auth0_Nonce_Handler::getInstance()->validate( $nonce ) ) {
+        throw new WP_Auth0_LoginFlowValidationException(
+          __( 'Invalid nonce', 'wp-auth0' )
+        );
       }
+    }
 
-      $decodedToken->user_id = $decodedToken->sub;
+    // Legacy userinfo property
+    $decodedToken->user_id = $decodedToken->sub;
 
-      if ( $this->login_user( $decodedToken, $token, null ) ) {
-        if ( ! empty( $state_decoded->interim ) ) {
-          include WPA0_PLUGIN_DIR . 'templates/login-interim.php';
-          exit();
-        } else {
-          if ( ! empty( $state_decoded->redirect_to ) && wp_login_url() !== $state_decoded->redirect_to ) {
-            $redirectURL = $state_decoded->redirect_to;
-          } else {
-            $redirectURL = $this->a0_options->get( 'default_login_redirection' );
-          }
+    if ( $this->login_user( $decodedToken, $token, null ) ) {
+
+      // Validated in $this->init_auth0()
+      $state_decoded = $this->get_state( TRUE );
 
-          wp_safe_redirect( $redirectURL );
-          exit;
+      if ( ! empty( $state_decoded->interim ) ) {
+        include WPA0_PLUGIN_DIR . 'templates/login-interim.php';
+      } else {
+        $redirect_to = ! empty( $state_decoded->redirect_to ) ? $state_decoded->redirect_to : null;
```

Comment:
Isn't this the same, or you need an explicit `null` later?

```php
$redirect_to = $state_decoded->redirect_to
```

Reply:
`$state_decoded->redirect_to` will throw an error if it does not exist, so I'm checking whether it's there or not with `empty()`